Allow the jitter entropy source to be used by the FIPS provider as an internal source

paulidale opened the pull request:
Doing so makes the FIPS provider non-compliant unless an entropy assessment and revalidation are also undertaken.

- [x] documentation is added or updated
- [x] tests are added or updated

t8m commented on CHANGES.md:
```diff
@@ -30,7 +30,13 @@ OpenSSL 3.4
- * none yet
+ * Optionally allow the FIPS provider to use the `JITTER` entropy source.
+   Note that using this option will require the resulting FIPS provider
+   to undergo validation by the [CMVP], without this the FIPS provider
```
```suggestion
   to undergo entropy source validation by the [CMVP], without this the FIPS provider
```

paulidale replied:
done.

t8m commented on crypto/rand/build.info:
```diff
@@ -11,5 +10,12 @@ IF[{- !$disabled{'deprecated-3.0'} -}]
+
 SOURCE[../../libcrypto]=$COMMON $CRYPTO
 SOURCE[../../providers/libfips.a]=$COMMON
+
```
Nit: please drop this extra line

paulidale replied:
& done

t8m reviewed (changes requested):
Just two nits

paulidale commented:
Comments addressed.

t8m reviewed (approved):
Approved, with the temporary commit dropped when merging

0140454 reviewed (commented):
There are confusions on viewing this PR, so I hope to get some guidance.

Please correct me if I am wrong. Thank you.

xnox commented on INSTALL.md:
```diff
@@ -841,6 +841,18 @@ Don't perform FIPS module Power On Self Tests.
+2. met the minimum required entropy as specified by [jitterentropy-library] and
+3. have had the resulting module FIPS validated by the [CMVP].
+
+Failure to do all of these will produce a non-compliant FIPS provider.
```
... And re-validate on every change/submission due to chaining. If on the other hand it is unchained like in https://github.com/openssl/openssl/pull/25275 then an existing certificate could be reused. Standalone static library has been validated before https://csrc.nist.gov/projects/cryptographic-module-validation-program/entropy-validations/certificate/191 that could be marked for reuse, with public downloads available.

Although self-submission of jitterentropy-library is straight forward.

paulidale replied:
As noted above, chaining is not happening inside the FIPS boundary.

paulidale replied:
I don't think certificate reuse is possible when changing platforms.  The quality of the entropy is strongly tied to the CPU and it's environment.  Any change there is going to throw the computation off and require a re-assessment.

Some changes might be hideable from NIST (e.g. upgrading main memory or changing a disc drive), but the reality is everything will impact the jitter entropy quality to some extent.

xnox replied:
Please mark as resolved

xnox commented on crypto/rand/rand_lib.c:
```diff
@@ -599,6 +601,10 @@ static EVP_RAND_CTX *rand_new_seed(OSSL_LIB_CTX *libctx)
     }
+# else /* !OPENSSL_NO_FIPS_JITTER */
+    name = "JITTER";
+    propq = "-fips";
```
But even a public one, if it was certified, could be marked as propq "fips=yes".

paulidale replied:
It really doesn't matter.  The fetch is from the internal library context which cannot be modified or altered by anything outside of the FIPS provider.  I'm just being careful and ignoring any global setting that might have been put in place, although I'm not sure how that would even happen.

0140454 replied:
Got it. Thank you.

xnox commented on crypto/rand/rand_lib.c:
```diff
@@ -754,7 +766,7 @@ EVP_RAND_CTX *RAND_get0_primary(OSSL_LIB_CTX *ctx)
     }
 
-#ifdef FIPS_MODULE
+#if defined(FIPS_MODULE) && defined(OPENSSL_NO_FIPS_JITTER)
```
Indeed, And it will safe on submissions too like mentioned in https://github.com/openssl/openssl/pull/25275

paulidale replied:
Inside the FIPS provider, #25415 replaces the primary DRBG with the health tests.  This means that, inside the FIPS provider, no DRBG is feeding another.

Outside the FIPS provider is pretty much irrelevant from NIST's perspective.  I kept the primary DRBG there because it greatly reduces the amount of entropy drawn from the system.

0140454 replied:
#25415 indeed replaces the primary DRBG with CRNGT via

```c
#ifdef FIPS_MODULE
    ret = rand_new_crngt(ctx, dgbl->seed);
#else
    ...
    ret = rand_new_drbg(ctx, dgbl->seed, PRIMARY_RESEED_INTERVAL,
                        PRIMARY_RESEED_TIME_INTERVAL, 1);
#endif
```

However, `#ifdef FIPS_MODULE` becomes `#if defined(FIPS_MODULE) && defined(OPENSSL_NO_FIPS_JITTER)` in this PR.

In the end, only FIPS provider built with `disable-fips-jitter` flag does this replacement.

Maybe we should keep `#ifdef FIPS_MODULE` here rather than change to `#if defined(FIPS_MODULE) ...`?

paulidale replied:
Good catch.  This isn't what I intended to do here.

xnox reviewed (commented):
Overall this is really nice, as it enables one to have just the fips provider using jitter entropy, rather than rely on top-level libcrypto. And it will enable to claim security strengths for generated keys.

The chaining is still problematic, it seems to add complexity for no gains; requiring additional submissions (as chain becomes part of conditioning, and creates need for ESV re-submissions all the time, rather than re-using existing ESV)

And anecdotally, unchained performance seems to be much better - but still need to create clean benchmarks for it. Even for non-fips case.

I will ping again about setting the ESV cert to mark it open for reuse. As then a certified prebuilt static library is available as a tarball download.

xnox commented:
BTW 3.4 is not released yet, and if you only want to make jitter entropy source be available as an internal source, we can remove build options that make it a top level one.

paulidale commented:
> BTW 3.4 is not released yet, and if you only want to make jitter entropy source be available as an internal source, we can remove build options that make it a top level one.

There is zero chance of this making 3.4 which is a bit annoying but wholly justifiable given its immaturity.  I certainly am not going to push for or even support its inclusion at this late stage.

We might be able to remove jitter as a public source with the justification that it will create ongoing legacy support.  However, I'm not all that convinced by that argument.  Jitter is useful outside of FIPS usage IMO.  One thing I did contemplate is making jitter the default source if it's been configured -- that seems sensible.  I doubt it would be accepted into 3.4 however.

paulidale commented on crypto/rand/rand_lib.c:
```diff
@@ -721,7 +733,7 @@ static EVP_RAND_CTX *rand_new_crngt(OSSL_LIB_CTX *libctx, EVP_RAND_CTX *parent)
     return ctx;
 }
-#endif
+#endif  /* FIPS_MODULE && !OPENSSL_NO_FIPS_JITTER) */
```
Will look into this tomorrow.  I suspect you are correct.  Too many negatives for my liking.

paulidale commented:
> And it will enable to claim security strengths for generated keys.

I am planning on introducing an _entropy quality_ measure in a future PR.  We've no real concept of _strength_ in OpenSSL currently and I'd really like to have _strength_ made pervasive throughout of the entire library.

paulidale commented:
Last two commits are the new changes, I squashed the rest up for tidiness's sake.

I fixed the [issue raised above](https://github.com/openssl/openssl/pull/25498#discussion_r1769721936).

Fixed a locking problem with the CRNG test filter.

Removed the `use_df = 0` for the public and private DRBGs was just wrong and breaks FIPS compliance because the derivation function is mandatory under FIPS.  This was introduced in 3.2 by #16156.  The fix therefore doesn't need back porting since it isn't impacting any version undergoing FIPS validation (if 3.4 is to be validated, the fix needs to go there).

paulidale commented:
Dropping the temporary run the daily tests commit.

Ready for re-review.

paulidale commented:
Thanks.

I've an open question about making jitter the default entropy source (outside of FIPS) if it is configured.
I'm also thinking about how to do entropy quality more nicely.

paulidale commented:
Failing check addressed by #25582 and are not related to the changes here.

paulidale commented:
Kicking CIs after #25582 has been merged.

paulidale commented:
Still awaiting reviews @openssl/committers

xnox commented:
I wonder how can I become `@openssl/committers` =)

slontis commented:
> I wonder how can I become `@openssl/committers` =)

A good way it to keep up the good work you are doing contributing to openssl.
https://openssl-library.org/policies/general/committer-policy

The requirements to maintain committer status have recently been updated in the bylaws
https://openssl-library.org/about/bylaws/index.html

slontis commented on Configure:
```diff
@@ -957,6 +960,11 @@ while (@argvcopy)
+                elsif ($1 eq "fips-jitter")
+                        {
+                        delete $disabled{"fips"};
+                        delete $disabled{"jitter"};
```
Is this right?
Could you potentially want to build fips provider with jitter but not use it in the default provider?

paulidale replied:
This just means that it is also available in the FIPS provider.  It won't be used unless config specifies it or the appropriate call is made to enable it.

I.e. this doesn't automatically make it _used_ in the default provider.

slontis replied:
Looking at build.info.. It does this

````
SUBDIRS=seeding

$RANDS_GOAL=../../libdefault.a ../../libfips.a

SOURCE[$RANDS_GOAL]=drbg.c test_rng.c drbg_ctr.c drbg_hash.c drbg_hmac.c
SOURCE[../../libdefault.a]=seed_src.c seed_src_jitter.c
SOURCE[../../libfips.a]=fips_crng_test.c
````

So the seed_src_jitter.c file is doing the check (which it probably shouldnt)..

The base and default provider both do this
````
static const OSSL_ALGORITHM deflt_rands[] = {
#ifndef OPENSSL_NO_JITTER
    { PROV_NAMES_JITTER, "provider=default", ossl_jitter_functions },
#endif
````

So enabling fips-jitter adds code to non fips source, which is what I disagree with.

paulidale replied:
Hmmmm.....  the other  jitter options also have to be specified to locate the jitter library and include directory and they don't have fips in the name.  Maybe I should document that fips-jitter also enables jitter or insist that both configure options are specified?  Having the code in the default provider is harmless since it won't be used but introduces a small amount of bloat.

Dissociating the two will be painful I suspect.

xnox replied:
possibly a non-issue, as most do not perform a single configure and build of default and fips providers. It is quite typical to configure one tree and install headers, libcrypto, libssl from it; and another one to just get fips.so from. meaning most often default provider / libcrypto is discarded in the fips builds. But having this cleaner would be nice. Because if one only intents to use fips.so, one really does not need jitter in libcrypto.

Separately, I have been pondering a no-default build..... which doesn't link libdefault.a into libcrypto, such that default libcrypto only has base provider. Specifically for builds that only care about fips.so and nothing else. As it makes it possible to provide libcrypto for fips-certified-only environments witout ability to programatically load and use '-fips/fips=no' algorithms.

paulidale replied:
Good point.  Once the first CVE fix comes out, split builds will be normal I suspect.

paulidale commented:
Essentially, it boils down to submitting pull requests and doing reviews.
Do enough quality ones of each and you'll be noticed.

At least that's the current situation but with the OTC disappearing at some point the policy will change.  I expect the requirements of reviews and PRs won't change significantly.

slontis commented on INSTALL.md:
```diff
@@ -841,6 +841,18 @@ Don't perform FIPS module Power On Self Tests.
+
+1. independently performed the required [SP 800-90B] entropy assessments;
+2. met the minimum required entropy as specified by [jitterentropy-library] and
+3. have had the resulting module FIPS validated by the [CMVP].
```
```suggestion
3. have had the resulting FIPS provider certified by the [CMVP].
```

slontis replied:
Never noticed the suggestion option until I just saw someone else using it..

slontis commented on NEWS.md:
```diff
@@ -37,7 +37,13 @@ changes:
 This release adds the following new features:
 
-  * none yet
+  * Allow the FIPS provider to use the optional `JITTER` seed source.
```
```suggestion
  * Allow the FIPS provider to optionally use the `JITTER` seed source.
```

slontis commented on README-FIPS.md:
```diff
@@ -167,6 +167,18 @@ manual page.
+jitter entropy source will be used instead.  Note that this will cause
+the FIPS provider to operate in a non-compliant mode unless an entropy
+assessment and validation through the [CMVP] are additionally conducted.
+
```
```suggestion
Note that that "enable-fips-jitter" option is only available for OpenSSL 3.5 or greater.
```

paulidale replied:
I'm not sure that this is required, this README will only be distributed in 3.5 and higher...

I guess it doesn't hurt.

paulidale replied:
I fixed the wording too :grin: 

> Note that the `enable-fips-jitter` option is only available in OpenSSL
versions 3.5 and later.

slontis commented on crypto/rand/rand_lib.c:
```diff
@@ -615,15 +621,21 @@ static EVP_RAND_CTX *rand_new_seed(OSSL_LIB_CTX *libctx)
 }
+#endif  /* !FIPS_MODULE || !OPENSSL_NO_FIPS_JITTER */
 
+#ifndef FIPS_MODULE
```
Adding a comment tag to the end on line #653 would help here :)

slontis commented on crypto/rand/rand_lib.c:
```diff
@@ -722,7 +734,7 @@ static EVP_RAND_CTX *rand_new_crngt(OSSL_LIB_CTX *libctx, EVP_RAND_CTX *parent)
     return ctx;
 }
-#endif
+#endif  /* FIPS_MODULE && OPENSSL_NO_FIPS_JITTER */
```
```suggestion
#endif  /* FIPS_MODULE */
```

slontis commented on crypto/rand/rand_lib.c:
```diff
@@ -755,14 +767,19 @@ EVP_RAND_CTX *RAND_get0_primary(OSSL_LIB_CTX *ctx)
-    ret = rand_new_crngt(ctx, dgbl->seed);
-#else
+#if !defined(FIPS_MODULE) || !defined(OPENSSL_NO_FIPS_JITTER)
+    /* Create a seed source for libcrypto or jitter enabled  FIPS provider */
```
extra space before FIPS

paulidale commented:
Feedback addressed.

slontis commented on doc/man7/EVP_RAND-JITTER.pod:
```diff
@@ -86,6 +86,12 @@ are additionally conducted.
 L<EVP_RAND(3)>,
 L<EVP_RAND(3)/PARAMETERS>
 
+=head1 HISTORY
```
Maybe just a NOTE would be better since the configuration options are not part of EVP_RAND-JITTER.pod

paulidale replied:
Changed.

paulidale commented:
Rebase & squash with zero changes.

openssl-machine commented:
This pull request is ready to merge

xnox commented on CHANGES.md:
```diff
@@ -30,7 +30,13 @@ OpenSSL 3.4
+   to undergo entropy source validation by the [CMVP], without this
+   the FIPS provider will not be FIPS compliant.  Enable this using the
+   configuration option `enable-fips-jitter`.
+
```
ESV certification is separate from CMVP certification. When submitting for CMVP certification, one should already obtain ESV certificate. The same ESV certificate can be used multiple times for multiple CMVP submissions. The way the jitter entropy APIs are used, and the way it is linked statically, it is possible to validate jitter entropy library as a static library, and reuse it for multiple submissions. See for example public use document of https://csrc.nist.gov/projects/cryptographic-module-validation-program/entropy-validations/certificate/191

So this caveat is more strict than it actually is.

The idea is that many/most vendors have jitter entropy certified, and with these changes they can reuse their builds straight away. As most vendors have ESV for jitterentropy in kernel/openssl/gcrypt today, and this change makes it possible to reuse ESV, rather than require resubmissions each time. Previously chaining was causing the ESV to be non-reusable, and needed to reobtain ESV for every OpenSSL build change, due to OpenSSL DRBG being in the ESV boundary; which it no-longer is thanks to static linking of the standalone build/library and unchaining.

xnox commented on INSTALL.md:
```diff
@@ -841,6 +841,18 @@ Don't perform FIPS module Power On Self Tests.
+2. met the minimum required entropy as specified by [jitterentropy-library] and
+3. have had the resulting FIPS provider certified by the [CMVP].
+
+Failure to do all of these will produce a non-compliant FIPS provider.
```
This is imprecise

```
Step 3 is Obtain ESV certificate for jitterentropy-library [ESV]
Step 4 have had the resulting FIPS provider certified by the [CMVP]

[ESV] https://csrc.nist.gov/Projects/cryptographic-module-validation-program/entropy-validations
```

Note that the same build/ESV can be used multiple times for different CMVP submissions of the cryptographic module, whilst compatible the same ESV cert can be reused over and over again.

xnox commented on NEWS.md:
```diff
@@ -37,7 +37,13 @@ changes:
+    Because this seed source is not part of the OpenSSL FIPS validations,
+    it should only be enabled after the [jitterentropy-library] has been
+    assessed for entropy quality.  Moreover, the FIPS provider including
+    this entropy source will need to be validated by the [CVMP] before
```
`will need to be validated by the [ESV] before`

xnox commented on README-FIPS.md:
```diff
@@ -167,6 +167,21 @@ manual page.
+enabling the `enable-fips-jitter` option during configuration, an internal
+jitter entropy source will be used instead.  Note that this will cause
+the FIPS provider to operate in a non-compliant mode unless an entropy
+assessment and validation through the [CMVP] are additionally conducted.
```
[ESV]

xnox commented on NEWS.md:
```diff
@@ -2007,3 +2013,5 @@ OpenSSL 0.9.x
 [README-QUIC.md]: ./README-QUIC.md
 [issue tracker]: https://github.com/openssl/openssl/issues
+[CMVP]: https://csrc.nist.gov/projects/cryptographic-module-validation-program
+[jitterentropy-library]: https://github.com/smuellerDD/jitterentropy-library
```
replace [CMVP] with [ESV] https://csrc.nist.gov/Projects/cryptographic-module-validation-program/entropy-validations

xnox commented on README-FIPS.md:
```diff
@@ -167,6 +167,21 @@ manual page.
+versions 3.5 and later.
+
+ [CMVP]: https://csrc.nist.gov/projects/cryptographic-module-validation-program
+
```
replace [CMVP] with [ESV] https://csrc.nist.gov/Projects/cryptographic-module-validation-program/entropy-validations

xnox commented on doc/man7/EVP_RAND-JITTER.pod:
```diff
@@ -44,6 +44,15 @@ A context for the seed source can be obtained by calling:
+provider will use an internal jitter source for its entropy.  Enabling
+this option will cause the FIPS provider to operate in a non-compliant
+mode unless an entropy assessment and validation through the
+L<CMVP|https://csrc.nist.gov/projects/cryptographic-module-validation-program>
```
[ESV] https://csrc.nist.gov/Projects/cryptographic-module-validation-program/entropy-validations

xnox commented:
For avoidance of any ambiguity, whenever talking about entropy validation, please refer to the "ESV" - "Entropy Source Validation" as it is a separate validation and certification programme, related to CMVP, but distinct from it.

As one has to complete and receive ESV first, before one can complete a CMVP submission. But then later one can do multiple CMVP submissions with the same ESV.

Note as the changes are documentation only; this PR can be merged as is. And separately I can submit a PR to correct the docs / news / changes / comments. As there is not need to invalidate the current reviews & CI results.

paulidale commented:
Amended references to CMVP to also mention ESV.

xnox reviewed (approved):
Also verified that only new changes pushed since previous approvals got dismissed are documentation only updates.

slontis commented on INSTALL.md:
```diff
@@ -841,6 +841,19 @@ Don't perform FIPS module Power On Self Tests.
+
+1. independently performed the required [SP 800-90B] entropy assessments;
+2. meet the minimum required entropy as specified by [jitterentropy-library];
+3. obtain an [ESV] certificate for jitterentropy-library and
```
```suggestion
3. obtain an [ESV] certificate for the |jitterentropy-library| and
```

paulidale replied:
Not sure we need the same link on sequential lines...

slontis reviewed (approved):
Apart from a small NIT it looks good to me..

paulidale commented:
Clean rebase and squash to recheck the CIs after recent merges.

paulidale commented:
CI failures are a result of #25632 and aren't relevant to this PR.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been  updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.
